packages/doenetml-worker/lib-doenetml-core/src/dast/ref_resolve/root_names.rs

```diff
@@ -0,0 +1,269 @@
+                            let resolutions = resolutions_to_node.get_mut(*res_idx).unwrap();
+                            let name_resolution =
+                                resolutions.by_name.get_or_insert(ResolutionViaName {
+                                    name: name.to_string(),
+                                    origins: Vec::new(),
+                                });
+
+                            if name_resolution.name != *name {
+                                panic!("Inconsistent resolver data as node can be reached by two different names");
+                            }
+
+                            name_resolution.origins.push(origin);
+                        }
+                        Ref::Ambiguous(_) => {}
+                    }
+                }
+            }
+            ResolutionAlgorithm::DontSearchChildren | ResolutionAlgorithm::Unsearchable => {}
+        }
+
+        for (index, resolution) in node_data.index_resolutions.iter().enumerate() {
+            if let Some(res_idx) = resolution {
+                // record the fact that `res_idx` is the resolution for a reference to `index` starting at `origin`.
+                let resolutions = resolutions_to_node.get_mut(*res_idx).unwrap();
+
+                if resolutions.by_index.is_some() {
+                    panic!("Inconsistent resolver data as node can be reached by two different indices");
+                }
+
+                resolutions.by_index = Some(ResolutionViaIndex { index, origin });
+            }
+        }
+    }
+
+    // By concatenating all the series of resolutions from which one can get to each node starting a the root,
+    // compile a list of all names by which a node can be referenced from the root.
+    // When complete all entries will contain a list `NamesReachableFromRoot`
+    let mut reachable_names: Vec<Option<NamesReachableFromRoot>> =
+        iter::repeat_n(None, n_nodes).collect::<Vec<_>>();
+    for idx in 0..n_nodes {
+        reachable_names = concatenate_reachable_names(idx, reachable_names, &resolutions_to_node);
+    }
+
+    // The final root name for each node is the "simplest" of the reachable names where "simplest" is defined by
+    // the fewest number of pieces (indices or sub-names) used to construct the name.
+    // Ties are broken by the fewest number of indices used, followed by the shortest overall name.
+    reachable_names
+        .into_iter()
+        .map(|possible_names| {
+            possible_names.unwrap().into_iter().reduce(|acc, name| {
+                let index_count_name = name.chars().filter(|c| *c == ':').count();
+                let index_count_acc = acc.chars().filter(|c| *c == ':').count();
+                let separator_count_name =
+                    index_count_name + name.chars().filter(|c| *c == '.').count();
+                let separator_count_acc =
+                    index_count_acc + acc.chars().filter(|c| *c == '.').count();
```

Comment:
Can you do this computation before you do `concatenate_reachable_names`? It seems a little strange to turn things to a string, erasing the data types, then re-guessing the data types based on the string.